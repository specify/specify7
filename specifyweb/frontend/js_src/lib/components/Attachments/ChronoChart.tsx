import React from 'react';

import { useBooleanState } from '../../hooks/useBooleanState';
import { commonText } from '../../localization/common';
import { headerText } from '../../localization/header';
import { Button } from '../Atoms/Button';
import { ImageViewer } from '../Attachments/ImageViewer';
import { Dialog, dialogClassNames } from '../Molecules/Dialog';

export function ChronoChart(): JSX.Element {
  const [showChronoChart, handleShowChronoChart, handleHideChronoChart] =
    useBooleanState();

  return (
    <>
      <Button.Icon
        icon="clock"
        title="Chronostratigraphic Chart"
        onClick={handleShowChronoChart}
      />
      {showChronoChart && (
        <Dialog
          buttons={
            <Button.DialogClose>{commonText.close()}</Button.DialogClose>
          }
          className={{
            container: dialogClassNames.wideContainer,
          }}
          header={headerText.chronostratigraphicChart()}
          onClose={handleHideChronoChart}
        >
          <ImageViewer
            alt="Chrono Chart"
            src="/static/img/chronostratChart2023-09.jpg"
          />
        </Dialog>
      )}
    </>
  );
}