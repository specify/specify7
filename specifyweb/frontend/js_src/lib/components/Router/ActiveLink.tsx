import type { Path } from '@remix-run/router';
import React from 'react';
import { useLocation } from 'react-router-dom';

import { Link } from '../Atoms/Link';
import { locationToUrl } from './queryString';

/**
 * Mark link that points to the current page
 * Based on react-router's <NavLink>
 */
export function ActiveLink<T extends Parameters<typeof Link.Default>[0]>({
  component: LinkComponent = Link.Default,
  'aria-current': ariaCurrent = 'page',
  mode = 'exact',
  activeOverride,
  className,
  ...props
}: Omit<T, 'className'> & {
  readonly mode?: 'exact' | 'startsWith';
  readonly activeOverride?: boolean;
  readonly component?: (props: T) => JSX.Element;
  readonly className?: string | ((isActive: boolean) => string);
}): JSX.Element {
  const rawIsActive = useIsActive(props.href, mode === 'exact');
  const isActive = activeOverride ?? rawIsActive;
  return (
    <LinkComponent
      {...(props as T)}
      aria-current={isActive ? ariaCurrent : undefined}
      className={
        typeof className === 'function' ? className(isActive) : className
      }
    />
  );
}

export function useIsActive(rawHref: string, isExact: boolean): boolean {
  const location = useLocation();
  return isSameUrl(location, rawHref, isExact);
}

export function isSameUrl(
  location: Path,
  path: string,
  isExact: boolean
): boolean {
  const pathName = location.pathname.endsWith('/')
    ? location.pathname
    : `${location.pathname}/`;
  const currentUrl = locationToUrl(location);
  const rawUrl = new URL(path, `${globalThis.location.origin}${currentUrl}`);
  const url = {
    pathName: rawUrl.pathname.endsWith('/')
      ? rawUrl.pathname
      : `${rawUrl.pathname}/`,
    hash: rawUrl.hash,
  };
  const fullUrl = `${url.pathName}${url.hash}`;
  return isExact
    ? fullUrl === `${pathName}${location.hash}`
    : pathName.startsWith(url.pathName);
}
