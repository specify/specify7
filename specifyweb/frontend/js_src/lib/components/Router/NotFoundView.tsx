import React from 'react';

import { mainText } from '../../localization/main';
import { Http } from '../../utils/ajax/definitions';
import { Container } from '../Atoms';
import { icons } from '../Atoms/Icons';
import { Link } from '../Atoms/Link';
import { useTitle } from '../Molecules/AppTitle';

export function NotFoundView({
  container = true,
}: {
  readonly container?: boolean;
}): JSX.Element {
  useTitle(mainText.pageNotFound());
  const content = (
    <Container.Center className="flex flex-1">
      <div className="flex flex-1 flex-col items-center justify-center gap-2 text-center">
        <h2
          aria-label={mainText.pageNotFound()}
          className="py-2 text-9xl text-indigo-300"
        >
          {Http.NOT_FOUND}
        </h2>
        <p>{mainText.nothingWasFound()}</p>
        <p>{mainText.pageNotFoundDescription()}</p>
        <Link.Default href="/specify/">
          {icons.arrowLeft}
          {mainText.returnToHomepage()}
        </Link.Default>
      </div>
    </Container.Center>
  );
  return container ? (
    <Container.FullGray>{content}</Container.FullGray>
  ) : (
    content
  );
}
