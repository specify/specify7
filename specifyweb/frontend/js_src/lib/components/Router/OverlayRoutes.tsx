import React from 'react';

import { attachmentsText } from '../../localization/attachments';
import { commonText } from '../../localization/common';
import { headerText } from '../../localization/header';
import { interactionsText } from '../../localization/interactions';
import { mergingText } from '../../localization/merging';
import { queryText } from '../../localization/query';
import { reportsText } from '../../localization/report';
import { schemaText } from '../../localization/schema';
import { treeText } from '../../localization/tree';
import { userText } from '../../localization/user';
import { welcomeText } from '../../localization/welcome';
import { wbText } from '../../localization/workbench';
import type { RA } from '../../utils/types';
import { Redirect } from './Redirect';
import type { EnhancedRoute } from './RouterUtils';

/* eslint-disable @typescript-eslint/promise-function-async */
/**
 * Overlay routes are rendered inside of a <Dialog>.
 * When linking to an overlay route, the previous route remains visible in
 * the background.
 * If opening overlay route in a new tab, the background is black.
 *
 * Overlay routes can also get a handleClose callback like this:
 *
 * ```js
 * const handleClose = React.useContext(OverlayContext);
 * ```
 *
 * Calling handleClose() would return user back to what was rendered behind
 * the overlay. If nothing was rendered behind the overlay, user is returned
 * to the Welcome Page
 */
export const overlayRoutes: RA<EnhancedRoute> = [
  {
    path: 'overlay',
    children: [
      {
        path: 'user-tools',
        title: headerText.userTools(),
        element: () =>
          import('../Header/UserTools').then(
            ({ UserToolsOverlay }) => UserToolsOverlay
          ),
      },
      {
        path: 'simple-search',
        title: headerText.simpleSearch(),
        element: () =>
          import('../Header/ExpressSearchTask').then(
            ({ ExpressSearchOverlay }) => ExpressSearchOverlay
          ),
      },
      {
        path: 'express-search',
        element: <Redirect to="/specify/overlay/simple-search/" />,
      },
      {
        path: 'choose-collection',
        title: commonText.chooseCollection(),
        element: () =>
          import('../Header/ChooseCollection').then(
            ({ ChooseCollection }) => ChooseCollection
          ),
      },
      {
        path: 'data-entry',
        title: headerText.dataEntry(),
        element: () =>
          import('../DataEntryTables').then(
            ({ FormsDialogOverlay }) => FormsDialogOverlay
          ),
      },
      {
        path: 'trees',
        title: treeText.trees(),
        element: () =>
          import('../Toolbar/TreeRepair').then(
            ({ TreeSelectOverlay }) => TreeSelectOverlay
          ),
      },
      {
        path: 'interactions',
        title: interactionsText.interactions(),
        children: [
          {
            index: true,
            element: () =>
              import('../Interactions/InteractionsDialog').then(
                ({ InteractionsOverlay }) => InteractionsOverlay
              ),
          },
          {
            path: 'return-loan',
            element: () =>
              import('../Interactions/InteractionsDialog').then(
                ({ InteractionLoanReturn }) => InteractionLoanReturn
              ),
          },
          {
            path: 'create/:tableName',
            element: () =>
              import('../Interactions/InteractionsDialog').then(
                ({ InteractionAction }) => InteractionAction
              ),
          },
        ],
      },
      {
        path: 'queries',
        title: queryText.queries(),
        element: () =>
          import('../Toolbar/Query').then(
            ({ QueriesOverlay }) => QueriesOverlay
          ),
        children: [
          {
            index: true,
            element: () =>
              import('../Toolbar/Query').then(
                ({ QueryListOutlet }) => QueryListOutlet
              ),
          },
          {
            path: 'new',
            title: queryText.newQueryName(),
            element: () =>
              import('../Toolbar/Query').then(({ NewQuery }) => NewQuery),
          },
        ],
      },
      {
        path: 'record-sets',
        title: commonText.recordSets(),
        element: () =>
          import('../Toolbar/RecordSets').then(
            ({ RecordSetsOverlay }) => RecordSetsOverlay
          ),
      },
      {
        path: 'reports',
        title: reportsText.reports(),
        element: () =>
          import('../Reports').then(({ ReportsOverlay }) => ReportsOverlay),
      },
      {
        path: 'data-sets',
        title: wbText.workBench(),
        element: () =>
          import('../Toolbar/WbsDialog').then(
            ({ DataSetsOverlay }) => DataSetsOverlay
          ),
      },
      {
        path: 'workbench/:dataSetId/meta',
        element: () =>
          import('../Toolbar/WbsDialog').then(
            ({ DataSetMetaOverlay }) => DataSetMetaOverlay
          ),
      },
      {
        path: 'tree-repair',
        title: headerText.repairTree(),
        element: () =>
          import('../Toolbar/TreeRepair').then(
            ({ TreeRepairOverlay }) => TreeRepairOverlay
          ),
      },
      {
        path: 'master-key',
        title: userText.generateMasterKey(),
        element: () =>
          import('../Toolbar/MasterKey').then(
            ({ MasterKeyOverlay }) => MasterKeyOverlay
          ),
      },
      {
        path: 'make-dwca',
        title: headerText.makeDwca(),
        element: () =>
          import('../ExportFeed/Dwca').then(
            ({ MakeDwcaOverlay }) => MakeDwcaOverlay
          ),
      },
      {
        path: 'force-update-feed',
        title: headerText.updateExportFeed(),
        element: () =>
          import('../ExportFeed/ForceUpdate').then(
            ({ ForceUpdateFeedOverlay }) => ForceUpdateFeedOverlay
          ),
      },
      {
        path: 'about',
        title: welcomeText.aboutSpecify(),
        element: () =>
          import('../HomePage/AboutSpecify').then(
            ({ AboutOverlay }) => AboutOverlay
          ),
      },

      {
        path: 'resources/app-resource/:id/*',
        element: () =>
          import('../AppResources/DialogEditor').then(
            ({ DialogEditor }) => DialogEditor
          ),
        isSingleResource: true,
      },
      {
        // This path is used when an overlay triggered a 404 page
        path: 'not-found',
        element: () =>
          import('../Router/RouterUtils').then(
            ({ NotFoundDialog }) => NotFoundDialog
          ),
      },
      {
        path: 'merge/:tableName',
        title: mergingText.mergeRecords(),
        element: () =>
          import('../Merging/index').then(({ MergingDialog }) => MergingDialog),
      },
      {
        path: 'attachments/import',
        title: attachmentsText.importAttachments(),
        element: () =>
          import('../AttachmentsBulkImport/Datasets').then(
            ({ AttachmentsImportOverlay }) => AttachmentsImportOverlay
          ),
      },
      {
        path: 'configure/uniqueness/:tableName',
        title: schemaText.uniquenessRules(),
        element: () =>
          import('../SchemaConfig/TableUniquenessRules').then(
            ({ TableUniquenessRules }) => TableUniquenessRules
          ),
      },
    ],
  },
];

/* eslint-enable @typescript-eslint/promise-function-async */
