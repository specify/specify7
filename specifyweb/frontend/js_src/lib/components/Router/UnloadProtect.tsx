import type { SafeLocation } from 'history';
import React from 'react';
import { unstable_useBlocker as useBlocker } from 'react-router';

import { commonText } from '../../localization/common';
import { mainText } from '../../localization/main';
import { f } from '../../utils/functools';
import type { GetOrSet, RA } from '../../utils/types';
import { Button } from '../Atoms/Button';
import { Dialog } from '../Molecules/Dialog';
import { locationToUrl } from './queryString';

export function UnloadProtectDialog({
  children,
  onCancel: handleCancel,
  onConfirm: handleConfirm,
}: {
  readonly children: string;
  readonly onCancel: () => void;
  readonly onConfirm: () => void;
}): JSX.Element {
  return (
    <Dialog
      buttons={
        <>
          <Button.DialogClose>{commonText.cancel()}</Button.DialogClose>
          <Button.Danger onClick={handleConfirm}>
            {mainText.leave()}
          </Button.Danger>
        </>
      }
      forceToTop
      header={mainText.leavePageConfirmation()}
      onClose={handleCancel}
    >
      {children}
    </Dialog>
  );
}

/**
 * List of current unload protects (used for preventing loss of unsaved changes)
 */
export const UnloadProtectsContext = React.createContext<
  RA<string> | undefined
>(undefined);
UnloadProtectsContext.displayName = 'UnloadProtectsContext';

export const UnloadProtectsRefContext = React.createContext<
  { readonly current: RA<string> } | undefined
>(undefined);
UnloadProtectsRefContext.displayName = 'UnloadProtectsRefContext';

export const SetUnloadProtectsContext = React.createContext<
  GetOrSet<RA<string>>[1] | undefined
>(undefined);
SetUnloadProtectsContext.displayName = 'SetUnloadProtectsContext';

/**
 * This component should only be used by the router
 *
 * Handles blocking navigation and displaying an unload protect dialog as
 * appropriate
 */
export function RouterUnloadProtect({
  singleResource,
}: {
  readonly singleResource: string | undefined;
}): JSX.Element | null {
  const unloadProtects = React.useContext(UnloadProtectsContext)!;
  const unloadProtectsRef = React.useContext(UnloadProtectsRefContext)!;

  const blocker = useBlocker(
    React.useCallback<Exclude<Parameters<typeof useBlocker>[0], boolean>>(
      ({ nextLocation, currentLocation }) =>
        unloadProtectsRef.current.length > 0 &&
        hasUnloadProtect(nextLocation, currentLocation, singleResource),
      [singleResource]
    )
  );

  // Remove the blocker if nothing is blocking
  const isEmpty = unloadProtects.length === 0;
  const isSet = blocker.state === 'blocked';
  const shouldUnset = isEmpty && isSet;
  React.useEffect(
    () => (shouldUnset ? blocker.proceed() : undefined),
    [isEmpty, isSet, blocker]
  );

  return (
    <>
      {blocker.state === 'blocked' && unloadProtects.length > 0 ? (
        <UnloadProtectDialog
          onCancel={(): void => blocker.reset()}
          onConfirm={(): void => blocker.proceed()}
        >
          {unloadProtects.at(-1)!}
        </UnloadProtectDialog>
      ) : null}
    </>
  );
}

/** Decide whether a given URL change should trigger unload protect */
const hasUnloadProtect = (
  nextLocation: SafeLocation,
  currentLocation: SafeLocation,
  singleResource: string | undefined
): boolean =>
  // Check for navigation within overlay
  !pathIsOverlay(nextLocation.pathname) &&
  // Check for navigation that only changes query string / hash
  !isCurrentUrl(nextLocation.pathname) &&
  // Check for exiting overlay
  f.maybe(
    currentLocation.state?.type === 'BackgroundLocation'
      ? currentLocation.state.location
      : undefined,
    locationToUrl
  ) !== locationToUrl(nextLocation) &&
  // Check for navigation within single resource
  !isSingleResource(nextLocation, singleResource);

// Don't trigger unload protect if only query string or hash changes
const isCurrentUrl = (relativeUrl: string): boolean =>
  new URL(relativeUrl, globalThis.location.origin).pathname ===
  globalThis.location.pathname;

const isSingleResource = (
  { pathname }: SafeLocation,
  singleResource: string | undefined
): boolean =>
  singleResource !== undefined &&
  pathname.startsWith(singleResource) &&
  globalThis.location.pathname.startsWith(singleResource);

export const pathIsOverlay = (relativeUrl: string): boolean =>
  relativeUrl.startsWith('/specify/overlay/');
