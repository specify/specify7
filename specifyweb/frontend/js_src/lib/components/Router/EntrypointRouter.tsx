import React from 'react';
import {
  createBrowserRouter,
  Navigate,
  RouterProvider,
} from 'react-router-dom';

import { commonText } from '../../localization/common';
import { userText } from '../../localization/user';
import type { RA } from '../../utils/types';
import { NotFoundView } from './NotFoundView';
import type { EnhancedRoute } from './RouterUtils';
import { toReactRoutes } from './RouterUtils';

/* eslint-disable @typescript-eslint/promise-function-async */
export const entrypointRoutes: RA<EnhancedRoute> = [
  {
    path: 'accounts',
    children: [
      {
        path: 'login',
        title: userText.logIn(),
        element: () => import('../Login').then(({ Login }) => Login),
      },
      {
        path: 'legacy_login',
        title: userText.logIn(),
        element: () => import('../Login').then(({ Login }) => Login),
      },
      {
        path: 'choose_collection',
        title: commonText.chooseCollection(),
        element: () =>
          import('../ChooseCollection').then(
            ({ ChooseCollection }) => ChooseCollection
          ),
      },
      {
        path: 'password_change',
        title: userText.changePassword(),
        element: () =>
          import('../PasswordChange').then(
            ({ PasswordChange }) => PasswordChange
          ),
      },
    ],
  },
  {
    path: 'specify/*',
    element: () =>
      import('../Core/ContextLoader').then(
        ({ ContextLoader }) => ContextLoader
      ),
  },
  // This should never be reached as back-end has a redict, but good to have it just in case
  {
    index: true,
    element: <Navigate to="/specify/" />,
  },
  {
    path: '*',
    element: <NotFoundView />,
  },
];
/* eslint-enable @typescript-eslint/promise-function-async */

const routes = toReactRoutes(entrypointRoutes, undefined, false);
const router = createBrowserRouter(routes);

export function EntrypointRouter(): JSX.Element {
  return <RouterProvider router={router} />;
}
