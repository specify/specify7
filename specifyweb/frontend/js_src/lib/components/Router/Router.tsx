import type { SafeLocation } from 'history';
import React from 'react';
import type { SafeNavigateFunction } from 'react-router';
import { useLocation, useNavigate, useRoutes } from 'react-router-dom';

import { useErrorContext } from '../../hooks/useErrorContext';
import { toLocalUrl } from '../../utils/ajax/helpers';
import { listen } from '../../utils/events';
import { setDevelopmentGlobal } from '../../utils/types';
import { error } from '../Errors/assert';
import { crash, softFail } from '../Errors/Crash';
import { ErrorBoundary } from '../Errors/ErrorBoundary';
import { userPreferences } from '../Preferences/userPreferences';
import { NotFoundView } from './NotFoundView';
import { overlayRoutes } from './OverlayRoutes';
import { locationToUrl } from './queryString';
import { toReactRoutes } from './RouterUtils';
import { routes } from './Routes';
import { pathIsOverlay, RouterUnloadProtect } from './UnloadProtect';

let unsafeNavigateFunction: SafeNavigateFunction | undefined;
export const unsafeNavigate = (
  ...parameters: Parameters<SafeNavigateFunction>
): void => unsafeNavigateFunction?.(...parameters);
let unsafeLocation: SafeLocation | undefined;

/**
 * Using this should be avoided when possible. Render <NotFoundView /> or
 * <NotFoundDialog /> instead.
 *
 * If error was triggered by an overlay, display 404 error in a dialog. Otherwise,
 * replace the main page with a 404.
 *
 * BUG: This assumes that if overlay is open, then it triggered the 404
 *  error, which is not always the case as the main content could be sending
 *  requests in the background too
 */
export function unsafeTriggerNotFound(): boolean {
  unsafeNavigateFunction?.(
    unsafeLocation === undefined
      ? '/specify/'
      : pathIsOverlay(locationToUrl(unsafeLocation))
      ? '/specify/overlay/not-found/'
      : unsafeLocation,
    {
      replace: true,
      state:
        typeof unsafeLocation === 'object' &&
        pathIsOverlay(locationToUrl(unsafeLocation))
          ? unsafeLocation.state
          : { type: 'NotFoundPage' },
    }
  );
  return typeof unsafeNavigateFunction === 'function';
}

const transformedRoutes = toReactRoutes(routes);
const transformedOverlays = toReactRoutes(overlayRoutes);

/**
 * Wrapper for React-Router with support for overlays and unload protect.
 *
 * Support for overlays was modeled based on this example:
 * https://reactrouter.com/docs/en/v6/examples/modal
 */
export function Router(): JSX.Element {
  /*
   * REFACTOR: replace usages of navigate with <a> where possible
   * REFACTOR: replace <Button> with <Link> where possible
   */
  const navigate = useNavigate();

  const location = useLocation();
  unsafeLocation = location;
  const state = location.state;

  /*
   * BUG: direct calls to navigate() with overlay URL don't set
   *    BackgroundLocation. This is a partial workaround, but it won't stand on
   *    page reload
   */
  const previousLocation = React.useRef(location);
  const rawBackground =
    state?.type === 'BackgroundLocation' ? state.location : undefined;
  const backgroundRef = React.useRef(rawBackground);
  if (!pathIsOverlay(location.pathname)) backgroundRef.current = undefined;
  else if (rawBackground !== undefined) backgroundRef.current = rawBackground;
  else if (!pathIsOverlay(previousLocation.current.pathname))
    backgroundRef.current = previousLocation.current;
  const background = backgroundRef.current;
  previousLocation.current = location;

  const isNotFoundState =
    state?.type === 'NotFoundPage' ||
    background?.state?.type === 'NotFoundPage';
  useErrorContext('location', location);

  React.useEffect(() => {
    unsafeNavigateFunction = navigate;
    setDevelopmentGlobal('_goTo', navigate);
    // If page was in 404 state and user reloaded it, then clear the 404 state
    if (isNotFoundState) navigate(locationToUrl(location), { replace: true });
    // Only ever execute this on initial render
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useLinkIntercept(background);

  const main =
    useRoutes(transformedRoutes, background ?? location) ?? undefined;

  const overlay = useRoutes(transformedOverlays, location) ?? undefined;

  const urlNotFound = main === undefined && overlay === undefined;

  const [singleResource, setSingleResource] = React.useState<
    string | undefined
  >(undefined);
  return (
    <SetSingleResourceContext.Provider value={setSingleResource}>
      <RouterUnloadProtect singleResource={singleResource} />
      <OverlayLocation.Provider
        value={pathIsOverlay(location.pathname) ? location : undefined}
      >
        {isNotFoundState || urlNotFound ? <NotFoundView /> : undefined}
        {typeof overlay === 'object' && (
          <Overlay background={background} overlay={overlay} />
        )}
        {
          /**
           * If in 404 state, don't render the main content.
           * Fixes https://github.com/specify/specify7/issues/3339
           */
          isNotFoundState ? undefined : main
        }
      </OverlayLocation.Provider>
    </SetSingleResourceContext.Provider>
  );
}

function useLinkIntercept(background: SafeLocation | undefined): void {
  const navigate = useNavigate();
  const location = useLocation();

  const resolvedLocation = React.useRef<SafeLocation>(background ?? location);
  React.useEffect(() => {
    resolvedLocation.current = background ?? location;
  }, [background, location]);

  React.useEffect(
    () =>
      listen(document.body, 'click', (event) => {
        const parsed = parseClickEvent(event);
        if (parsed === undefined) return;
        const { url, isOverlay } = parsed;
        navigate(
          url,
          isOverlay
            ? {
                state: {
                  type: 'BackgroundLocation',
                  location: resolvedLocation.current,
                },
              }
            : undefined
        );
      }),
    [navigate]
  );
}

/* Partially inspired by react-router's useLinkClickHandler() */
function parseClickEvent(
  event: Readonly<MouseEvent>
): { readonly url: string; readonly isOverlay: boolean } | undefined {
  const link = (event.target as HTMLElement)?.closest('a');
  if (
    // Check if link already has an onClick that called event.preventDefault()
    !event.defaultPrevented &&
    link !== null &&
    link.getAttribute('href')?.startsWith('/specify/') === true &&
    link.getAttribute('download') === null &&
    !event.metaKey &&
    !event.shiftKey &&
    !event.ctrlKey &&
    (link.target === '' ||
      link.target === '_self' ||
      (event.altKey &&
        userPreferences.get('general', 'behavior', 'altClickToSupressNewTab')))
  ) {
    // Don't handle absolute URLs that lead to a different origin
    const localUrl = toLocalUrl(link.href);
    if (typeof localUrl === 'string') {
      event.preventDefault();
      if (
        process.env.NODE_ENV !== 'production' &&
        link.getAttribute('href')!.startsWith('.')
      )
        crash(
          new Error(
            'Relative URLs are not supported as they are unpredictable. ' +
              'Relative URL leads to different path depending on whether ' +
              'current URL has trailing slash or not. Consider calling ' +
              'resolveRelative() first'
          )
        );
      return {
        url: localUrl,
        isOverlay: pathIsOverlay(localUrl),
      };
    }
  }
  return undefined;
}

function Overlay({
  overlay,
  background,
}: {
  readonly overlay: JSX.Element | undefined;
  // This happens when user opened the overlay directly by going to the URL
  readonly background: SafeLocation | undefined;
}): JSX.Element {
  const navigate = useNavigate();

  function handleClose(): void {
    const backgroundUrl =
      typeof background === 'object' ? locationToUrl(background) : '/specify/';
    navigate(backgroundUrl, { state: background?.state });
  }

  const handleCloseRef = React.useRef(handleClose);
  handleCloseRef.current = handleClose;

  const handleCloseOverlay = React.useCallback(
    // False positive
    // eslint-disable-next-line functional/prefer-tacit
    () => handleCloseRef.current(),
    []
  );

  return (
    <OverlayContext.Provider value={handleCloseOverlay}>
      <ErrorBoundary dismissible>{overlay}</ErrorBoundary>
    </OverlayContext.Provider>
  );
}

function defaultOverlayContext(): void {
  softFail(new Error('Tried to close Overlay outside of an overlay'));
}

export const isOverlay = (overlayContext: () => void): boolean =>
  overlayContext !== defaultOverlayContext;

/**
 * When in overlay, this context provides a function that closes the overlay.
 */
export const OverlayContext = React.createContext<() => void>(
  defaultOverlayContext
);
OverlayContext.displayName = 'OverlayContext';

/**
 * Regardless of whether component is in overlay or not, if any overlay is open,
 * this will provide location of that component
 */
export const OverlayLocation = React.createContext<SafeLocation | undefined>(
  undefined
);
OverlayLocation.displayName = 'OverlayLocation';

/**
 * If set, links within this path won't trigger unload protection
 */
export const SetSingleResourceContext = React.createContext<
  (path: string | undefined) => void
>(() => error('SetSingleResourceContext is not defined'));
SetSingleResourceContext.displayName = 'SetSingleResourceContext';

export const exportsForTests = { parseClickEvent };
