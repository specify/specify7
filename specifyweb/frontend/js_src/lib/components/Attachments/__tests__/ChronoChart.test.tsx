import React from 'react';

import { mount } from '../../../tests/reactUtils';
import { UnloadProtectsContext } from '../../Router/UnloadProtect';
import { ChronoChart } from '../ChronoChart';

describe('ChronoChart', () => {
  test('simple render', () => {
    const { asFragment } = mount(<ChronoChart />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('dialog open and close', async () => {
    const { getAllByRole, user } = mount(
      <UnloadProtectsContext.Provider value={[]}>
        <ChronoChart />
      </UnloadProtectsContext.Provider>
    );

    const button = getAllByRole('button')[0];

    await user.click(button);

    const dialog = getAllByRole('dialog')[0];

    expect(dialog).toMatchSnapshot();

    const closeButton = getAllByRole('button')[3];

    await user.click(closeButton);

    expect(() => getAllByRole('dialog')).toThrow();
  });
});
