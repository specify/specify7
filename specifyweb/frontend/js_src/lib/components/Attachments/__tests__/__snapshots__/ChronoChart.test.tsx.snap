// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChronoChart dialog open and close 1`] = `
<div
  aria-describedby="modal-0-content"
  aria-labelledby="modal-0-header"
  aria-modal="true"
  class="ReactModal__Content 
        flex flex-col gap-2 p-4 outline-none rounded resize max-w-[90%] shadow-lg shadow-gray-500 max-h-screen min-w-[min(40rem,90%)]
        overflow-x-hidden text-neutral-900 duration-0
        dark:border dark:border-neutral-700 dark:text-neutral-200
        
        
    bg-gradient-to-bl from-gray-200 via-white
    to-white dark:from-neutral-800 dark:via-neutral-900 dark:to-neutral-900
  
      "
  role="dialog"
  style="transform: translate(0px,0px);"
  tabindex="-1"
>
  <div
    class="
          flex items-center gap-2 md:gap-4
          -m-4 cursor-move p-4
          flex-wrap
        "
    id="modal-0-handle"
  >
    <div
      class="flex items-center gap-2"
    >
      <h2
        class="font-semibold text-black dark:text-white text-xl"
        id="modal-0-header"
      >
        Chronostratigraphic Chart
      </h2>
    </div>
  </div>
  <div
    class="
            dark:text-neutral-350 -mx-1 flex-1 overflow-y-auto px-1 py-4
            text-gray-700 flex flex-col gap-2
          "
    id="modal-0-content"
  >
    <div
      class="flex flex-col items-center justify-center h-full w-full"
      style="--transition-duration: 0;"
    >
      <div
        class="react-transform-wrapper transform-component-module_wrapper__SPB86 "
      >
        <div
          class="react-transform-component transform-component-module_content__FBWxo "
          style="transform: translate(0px, 0px) scale(1);"
        >
          <img
            alt="Chrono Chart"
            class="max-w-full max-h-[70vh] object-contain"
            src="/static/img/chronostratChart2023-09.jpg"
          />
        </div>
      </div>
      <div
        class="flex gap-2 mt-4"
      >
        <button
          aria-label="Zoom"
          class="icon link rounded"
          title="Zoom"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="w-6 h-6 flex-shrink-0"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          aria-label="Unzoom"
          class="icon link rounded"
          title="Unzoom"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="w-6 h-6 flex-shrink-0"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          aria-label="Reset"
          class="icon link rounded"
          title="Reset"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="w-6 h-6 flex-shrink-0"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M15.312 11.424a5.5 5.5 0 0 1-9.201 2.466l-.312-.311h2.433a.75.75 0 0 0 0-1.5H3.989a.75.75 0 0 0-.75.75v4.242a.75.75 0 0 0 1.5 0v-2.43l.31.31a7 7 0 0 0 11.712-3.138.75.75 0 0 0-1.449-.39Zm1.23-3.723a.75.75 0 0 0 .219-.53V2.929a.75.75 0 0 0-1.5 0V5.36l-.31-.31A7 7 0 0 0 3.239 8.188a.75.75 0 1 0 1.448.389A5.5 5.5 0 0 1 13.89 6.11l.311.31h-2.432a.75.75 0 0 0 0 1.5h4.243a.75.75 0 0 0 .53-.219Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="flex gap-2 justify-end"
  >
    <button
      class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm button hover:brightness-90 dark:hover:brightness-125 bg-[color:var(--secondary-button-color)] text-gray-800 shadow-sm
dark:text-gray-100"
      type="button"
    >
      Close
    </button>
  </div>
</div>
`;

exports[`ChronoChart simple render 1`] = `
<DocumentFragment>
  <button
    aria-label="Chronostratigraphic Chart"
    class="icon link rounded"
    title="Chronostratigraphic Chart"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M10 18a8 8 0 100-16 8 8 0 000 16zm1-12a1 1 0 10-2 0v4a1 1 0 00.293.707l2.828 2.829a1 1 0 101.415-1.415L11 9.586V6z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;
